lib/ignore/ignore.go

```diff
@@ -203,11 +208,17 @@ func (m *Matcher) parseLocked(r io.Reader, file string) error {
 	}
 
 	m.skipIgnoredDirs = true
+	var previous string
 	for _, p := range patterns {
+		// We automatically add patterns with a /** suffix - no need to check them
+		if l := len(p.pattern); l > 3 && p.pattern[:len(p.pattern)-3] == previous {
+			continue
+		}
```

Comment:
This optimization adds confusion and has questionable performance gains.

Reply:
The comment is confusing: This is not an optimisation, this is needed for the case outlined in https://github.com/syncthing/syncthing/pull/6151#discussion_r346215210:

A user entering a path like `!/foo/b*/**` means we should descend into ignored dirs (to catch e.g. `foo/bar/baz`). Contrary a user entering `!/foo/b*` doesn't require us to descend, however we internally add an additional pattern `!/foo/b*/**` to catch everything below. And here because of the previous pattern there's no need to descend into ignored dirs.